src/github.com/matrix-org/dendrite/federationapi/writers/threepid.go

```diff
@@ -0,0 +1,195 @@
+	inv invite,
+) (*gomatrixserverlib.Event, error) {
+	// Build the event
+	builder := &gomatrixserverlib.EventBuilder{
+		Type:     "m.room.member",
+		Sender:   inv.Sender,
+		RoomID:   inv.RoomID,
+		StateKey: &inv.MXID,
+	}
+
+	content := common.MemberContent{
+		// TODO: Load the profile
+		Membership: "invite",
+		ThirdPartyInvite: &common.TPInvite{
+			Signed: inv.Signed,
+		},
+	}
+
+	if err := builder.SetContent(content); err != nil {
+		return nil, err
+	}
+
+	eventsNeeded, err := gomatrixserverlib.StateNeededForEventBuilder(builder)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ask the roomserver for information about this room
+	queryReq := api.QueryLatestEventsAndStateRequest{
+		RoomID:       builder.RoomID,
+		StateToFetch: eventsNeeded.Tuples(),
+	}
+	var queryRes api.QueryLatestEventsAndStateResponse
+	if err = queryAPI.QueryLatestEventsAndState(&queryReq, &queryRes); err != nil {
+		return nil, err
+	}
+
+	if !queryRes.RoomExists {
+		// TODO: Use federation to auth the event
+		return nil, nil
+	}
+
+	if err = fillDisplayName(builder, content, queryRes.StateEvents); err != nil {
+		return nil, err
+	}
+
+	// Finish building the event
+	builder.Depth = queryRes.Depth
+	builder.PrevEvents = queryRes.LatestEvents
+
+	authEvents := gomatrixserverlib.NewAuthEvents(nil)
+
+	for i := range queryRes.StateEvents {
+		authEvents.AddEvent(&queryRes.StateEvents[i])
+	}
+
+	refs, err := eventsNeeded.AuthEventReferences(&authEvents)
+	if err != nil {
+		return nil, err
+	}
+	builder.AuthEvents = refs
+
+	eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.Matrix.ServerName)
+	now := time.Now()
+	event, err := builder.Build(eventID, now, cfg.Matrix.ServerName, cfg.Matrix.KeyID, cfg.Matrix.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &event, nil
+}
+
+// fillDisplayName looks in the room's state events for a m.room.third_party_invite
+// event with the state key matching a given m.room.member event's content's token.
+// If such an event is found, fills the "display_name" attribute of the
+// "third_party_invite" structure in the m.room.member event with the display_name
+// from the m.room.third_party_invite event.
+// Returns an error if there was a problem parsing the m.room.third_party_invite
+// event's content or updating the m.room.member event's content.
+// Returns nil if no m.room.third_party_invite with a matching token could be
+// found. Returning an error isn't necessary in this case as the event will be
+// rejected by gomatrixserverlib.
+func fillDisplayName(
+	builder *gomatrixserverlib.EventBuilder, content common.MemberContent,
+	stateEvents []gomatrixserverlib.Event,
+) error {
+	// Look for the m.room.third_party_invite event
+	var thirdPartyInviteEvent *gomatrixserverlib.Event
+	for _, event := range stateEvents {
+		if event.Type() == "m.room.third_party_invite" && *(event.StateKey()) == content.ThirdPartyInvite.Signed.Token {
+			thirdPartyInviteEvent = &event
+		}
+	}
```

Comment:
If you pass in the AuthEvents object rather than the stateevents you can use https://godoc.org/github.com/matrix-org/gomatrixserverlib#AuthEvents.ThirdPartyInvite to get the 
third_party_invite.

Reply:
Indeed. I did this part before I knew how to correctly use `AuthEvents`, and didn't rezrite it afterwards.

Reply:
Done https://github.com/matrix-org/dendrite/pull/218/commits/eb5abbd96a97868a054b589f75d4e0b5e02c7d02

Reply:
Done https://github.com/matrix-org/dendrite/pull/218/commits/667408a8a444afe8eaef8ab53e0167b1606ca43d